File: backend/src/main/java/com/yourcompany/app/application/service/ProductService.java
package com.yourcompany.app.application.service;

import com.yourcompany.app.domain.model.Product;
import com.yourcompany.app.domain.model.Shop;
import com.yourcompany.app.domain.repository.ProductRepository;
import com.yourcompany.app.domain.repository.ShopRepository;
import com.yourcompany.app.application.dto.ProductCreateDto;
import com.yourcompany.app.application.dto.ProductResponseDto;
import com.yourcompany.app.application.dto.ProductUpdateDto;
import com.yourcompany.app.common.exception.BadRequestException;
import com.yourcompany.app.common.exception.ResourceNotFoundException;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
public class ProductService {

    private final ProductRepository productRepository;
    private final ShopRepository shopRepository;

    @Autowired
    public ProductService(ProductRepository productRepository, ShopRepository shopRepository) {
        this.productRepository = productRepository;
        this.shopRepository = shopRepository;
    }

    /**
     * Create a new product
     * @param productDto Data transfer containing product info.
     * @return ResponseDto created with shop details.
     */
    @Transactional
    public ProductResponseDto createProduct(ProductCreateDto productDto) {
        // Shop validation
        Shop shop = shopRepository.findById(productDto.getShopId())
            .orElseThrow(() -> new ResourceNotFoundException("Shop not found with id: " + productDto.getShopId()));
        
        // Convert DTO to entity
        Product product = productDto.toEntity(shop);
        
        // Set default values if needed
        if (product.getQuantity() == null) {
            product.setQuantity(0);
        }
        
        if (product.getDiscount() == null) {
            product.setDiscount(BigDecimal.ZERO);
        }
        
        // Save the product
        Product savedProduct = productRepository.save(product);
        
        // Return the response DTO
        return new ProductResponseDto(savedProduct);
    }
    
    /**
     * Get all products with optional pagination
     * @param pageable Pagination information
     * @return Page of product response DTOs
     */
    @Transactional(readOnly = true)
    public Page<ProductResponseDto> getAllProducts(Pageable pageable) {
        return productRepository.findAll(pageable)
                .map(ProductResponseDto::new);
    }
    
    /**
     * Get all products for a specific shop
     * @param shopId Shop identifier
     * @return List of product response DTOs
     */
    @Transactional(readOnly = true)
    public List<ProductResponseDto> getProductsByShopId(UUID shopId) {
        Shop shop = shopRepository.findById(shopId)
                .orElseThrow(() -> new ResourceNotFoundException("Shop not found with id: " + shopId));
        
        return productRepository.findByShop(shop).stream()
                .map(ProductResponseDto::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Get all active products
     * @return List of active product response DTOs
     */
    @Transactional(readOnly = true)
    public List<ProductResponseDto> getActiveProducts() {
        return productRepository.findByActiveTrue().stream()
                .map(ProductResponseDto::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Get all active products for a specific shop
     * @param shopId Shop identifier
     * @return List of active product response DTOs for the shop
     */
    @Transactional(readOnly = true)
    public List<ProductResponseDto> getActiveProductsByShopId(UUID shopId) {
        Shop shop = shopRepository.findById(shopId)
                .orElseThrow(() -> new ResourceNotFoundException("Shop not found with id: " + shopId));
        
        return productRepository.findByShopAndActiveTrue(shop).stream()
                .map(ProductResponseDto::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Get products by category
     * @param category Product category
     * @return List of product response DTOs
     */
    @Transactional(readOnly = true)
    public List<ProductResponseDto> getProductsByCategory(String category) {
        return productRepository.findByCategory(category).stream()
                .map(ProductResponseDto::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Get products by category for a specific shop
     * @param shopId Shop identifier
     * @param category Product category
     * @return List of product response DTOs
     */
    @Transactional(readOnly = true)
    public List<ProductResponseDto> getProductsByShopAndCategory(UUID shopId, String category) {
        Shop shop = shopRepository.findById(shopId)
                .orElseThrow(() -> new ResourceNotFoundException("Shop not found with id: " + shopId));
        
        return productRepository.findByShopAndCategory(shop, category).stream()
                .map(ProductResponseDto::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Get active products by category for a specific shop
     * @param shopId Shop identifier
     * @param category Product category
     * @return List of active product response DTOs
     */
    @Transactional(readOnly = true)
    public List<ProductResponseDto> getActiveProductsByShopAndCategory(UUID shopId, String category) {
        Shop shop = shopRepository.findById(shopId)
                .orElseThrow(() -> new ResourceNotFoundException("Shop not found with id: " + shopId));
        
        return productRepository.findByShopAndCategoryAndActiveTrue(shop, category).stream()
                .map(ProductResponseDto::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Get featured products
     * @return List of featured product response DTOs
     */
    @Transactional(readOnly = true)
    public List<ProductResponseDto> getFeaturedProducts() {
        return productRepository.findByIsFeaturedTrue().stream()
                .map(ProductResponseDto::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Get featured products for a specific shop
     * @param shopId Shop identifier
     * @return List of featured product response DTOs for the shop
     */
    @Transactional(readOnly = true)
    public List<ProductResponseDto> getFeaturedProductsByShopId(UUID shopId) {
        Shop shop = shopRepository.findById(shopId)
                .orElseThrow(() -> new ResourceNotFoundException("Shop not found with id: " + shopId));
        
        return productRepository.findByShopAndIsFeaturedTrue(shop).stream()
                .map(ProductResponseDto::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Get active featured products for a specific shop
     * @param shopId Shop identifier
     * @return List of active featured product response DTOs for the shop
     */
    @Transactional(readOnly = true)
    public List<ProductResponseDto> getActiveFeaturedProductsByShopId(UUID shopId) {
        Shop shop = shopRepository.findById(shopId)
                .orElseThrow(() -> new ResourceNotFoundException("Shop not found with id: " + shopId));
        
        return productRepository.findByShopAndIsFeaturedTrueAndActiveTrue(shop).stream()
                .map(ProductResponseDto::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Get a product by ID
     * @param id Product identifier
     * @return Product response DTO
     */
    @Transactional(readOnly = true)
    public ProductResponseDto getProductById(UUID id) {
        Product product = productRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Product not found with id: " + id));
        
        return new ProductResponseDto(product);
    }
    
    /**
     * Update a product
     * @param id Product identifier
     * @param updateDto Data for updating the product
     * @return Updated product response DTO
     */
    @Transactional
    public ProductResponseDto updateProduct(UUID id, ProductUpdateDto updateDto) {
        // Find the product
        Product product = productRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Product not found with id: " + id));
        
        // Ensure ID in DTO matches path parameter if provided
        if (updateDto.getId() != null && !updateDto.getId().equals(id)) {
            throw new BadRequestException("ID in request body must match path parameter");
        }
        
        // Update entity from DTO
        updateDto.updateEntity(product);
        
        // Save updated product
        Product updatedProduct = productRepository.save(product);
        
        // Return updated response DTO
        return new ProductResponseDto(updatedProduct);
    }
    
    /**
     * Update product stock quantity
     * @param id Product identifier
     * @param quantity New quantity
     * @return Updated product response DTO
     */
    @Transactional
    public ProductResponseDto updateProductQuantity(UUID id, Integer quantity) {
        if (quantity < 0) {
            throw new BadRequestException("Quantity cannot be negative");
        }
        
        Product product = productRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Product not found with id: " + id));
        
        product.setQuantity(quantity);
        Product updatedProduct = productRepository.save(product);
        
        return new ProductResponseDto(updatedProduct);
    }
    
    /**
     * Toggle product featured status
     * @param id Product identifier
     * @return Updated product response DTO
     */
    @Transactional
    public ProductResponseDto toggleFeaturedStatus(UUID id) {
        Product product = productRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Product not found with id: " + id));
        
        product.setIsFeatured(!product.getIsFeatured());
        Product updatedProduct = productRepository.save(product);
        
        return new ProductResponseDto(updatedProduct);
    }
    
    /**
     * Toggle product active status
     * @param id Product identifier
     * @return Updated product response DTO
     */
    @Transactional
    public ProductResponseDto toggleActiveStatus(UUID id) {
        Product product = productRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Product not found with id: " + id));
        
        product.setIsActive(!product.getIsActive());
        Product updatedProduct = productRepository.save(product);
        
        return new ProductResponseDto(updatedProduct);
    }
    
    /**
     * Delete a product
     * @param id Product identifier
     */
    @Transactional
    public void deleteProduct(UUID id) {
        if (!productRepository.existsById(id)) {
            throw new ResourceNotFoundException("Product not found with id: " + id);
        }
        
        productRepository.deleteById(id);
    }
    
    /**
     * Search products by name
     * @param keyword Search keyword
     * @return List of matching product response DTOs
     */
    @Transactional(readOnly = true)
    public List<ProductResponseDto> searchProductsByName(String keyword) {
        return productRepository.findByNameContainingIgnoreCase(keyword).stream()
                .map(ProductResponseDto::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Advanced search for products by name, description or category
     * @param keyword Search keyword
     * @return List of matching product response DTOs
     */
    @Transactional(readOnly = true)
    public List<ProductResponseDto> searchProducts(String keyword) {
        return productRepository.searchProducts(keyword).stream()
                .map(ProductResponseDto::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Search products within a specific shop
     * @param shopId Shop identifier
     * @param keyword Search keyword
     * @return List of matching product response DTOs
     */
    @Transactional(readOnly = true)
    public List<ProductResponseDto> searchProductsByShop(UUID shopId, String keyword) {
        Shop shop = shopRepository.findById(shopId)
                .orElseThrow(() -> new ResourceNotFoundException("Shop not found with id: " + shopId));
        
        return productRepository.searchProductsByShop(shop, keyword).stream()
                .map(ProductResponseDto::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Get products with available stock (quantity > 0)
     * @param minQuantity Minimum quantity threshold
     * @return List of product response DTOs
     */
    @Transactional(readOnly = true)
    public List<ProductResponseDto> getProductsWithStock(Integer minQuantity) {
        return productRepository.findByQuantityGreaterThan(minQuantity).stream()
                .map(ProductResponseDto::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Get products within a price range
     * @param minPrice Minimum price
     * @param maxPrice Maximum price
     * @return List of product response DTOs
     */
    @Transactional(readOnly = true)
    public List<ProductResponseDto> getProductsByPriceRange(BigDecimal minPrice, BigDecimal maxPrice) {
        return productRepository.findByPriceBetween(minPrice, maxPrice).stream()
                .map(ProductResponseDto::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Get products within a price range and specific category
     * @param minPrice Minimum price
     * @param maxPrice Maximum price
     * @param category Product category
     * @return List of product response DTOs
     */
    @Transactional(readOnly = true)
    public List<ProductResponseDto> getProductsByPriceRangeAndCategory(BigDecimal minPrice, BigDecimal maxPrice, String category) {
        return productRepository.findByPriceBetweenAndCategory(minPrice, maxPrice, category).stream()
                .map(ProductResponseDto::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Get products with discounts
     * @return List of product response DTOs with discounts
     */
    @Transactional(readOnly = true)
    public List<ProductResponseDto> getProductsWithDiscount() {
        return productRepository.findProductsWithDiscount().stream()
                .map(ProductResponseDto::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Get products with discounts for a specific shop
     * @param shopId Shop identifier
     * @return List of product response DTOs with discounts for the shop
     */
    @Transactional(readOnly = true)
    public List<ProductResponseDto> getProductsWithDiscountByShop(UUID shopId) {
        Shop shop = shopRepository.findById(shopId)
                .orElseThrow(() -> new ResourceNotFoundException("Shop not found with id: " + shopId));
        
        return productRepository.findProductsWithDiscountByShop(shop).stream()
                .map(ProductResponseDto::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Get products with minimum discount percentage
     * @param discountPercentage Minimum discount percentage
     * @return List of product response DTOs
     */
    @Transactional(readOnly = true)
    public List<ProductResponseDto> getProductsWithMinimumDiscountPercentage(Double discountPercentage) {
        return productRepository.findProductsWithMinimumDiscountPercentage(discountPercentage).stream()
                .map(ProductResponseDto::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Get low stock products
     * @param threshold Low stock threshold
     * @return List of product response DTOs with low stock
     */
    @Transactional(readOnly = true)
    public List<ProductResponseDto> getLowStockProducts(Integer threshold) {
        return productRepository.findLowStockProducts(threshold).stream()
                .map(ProductResponseDto::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Get low stock products for a specific shop
     * @param shopId Shop identifier
     * @param threshold Low stock threshold
     * @return List of product response DTOs with low stock for the shop
     */
    @Transactional(readOnly = true)
    public List<ProductResponseDto> getLowStockProductsByShop(UUID shopId, Integer threshold) {
        Shop shop = shopRepository.findById(shopId)
                .orElseThrow(() -> new ResourceNotFoundException("Shop not found with id: " + shopId));
        
        return productRepository.findLowStockProductsByShop(shop, threshold).stream()
                .map(ProductResponseDto::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Get products with upcoming stock depletion
     * @param threshold Low stock threshold
     * @return List of product response DTOs with upcoming stock depletion
     */
    @Transactional(readOnly = true)
    public List<ProductResponseDto> getProductsWithUpcomingStockDepletion(Integer threshold) {
        return productRepository.findProductsWithUpcomingStockDepletion(threshold).stream()
                .map(ProductResponseDto::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Get recently added products
     * @param pageable Pagination information
     * @return List of recently added product response DTOs
     */
    @Transactional(readOnly = true)
    public List<ProductResponseDto> getRecentProducts(Pageable pageable) {
        return productRepository.findRecentProducts(pageable).stream()
                .map(ProductResponseDto::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Get best selling products
     * @param pageable Pagination information
     * @return List of best selling product response DTOs
     */
    @Transactional(readOnly = true)
    public List<ProductResponseDto> getBestSellingProducts(Pageable pageable) {
        return productRepository.findBestSellingProducts(pageable).stream()
                .map(ProductResponseDto::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Get products updated since a specific date
     * @param date Reference date
     * @return List of product response DTOs updated since the specified date
     */
    @Transactional(readOnly = true)
    public List<ProductResponseDto> getProductsUpdatedSince(LocalDateTime date) {
        return productRepository.findByUpdatedAtAfter(date).stream()
                .map(ProductResponseDto::new)
                .collect(Collectors.toList());
    }
}

File: frontend/src/app/core/services/product/product.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Product, ProductCreateDto, ProductUpdateDto } from '../../../models/product.model';
import { environment } from '../../../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class ProductService {
  private apiUrl = `${environment.apiUrl}/api/products`;

  constructor(private http: HttpClient) {}

  // Create a new product
  createProduct(product: ProductCreateDto): Observable<Product> {
    console.log('Product-Service:== creating new product with', product);
    return this.http.post<Product>(this.apiUrl, product);
  }

  // Get a product by ID
  getProductById(productId: string): Observable<Product> {
    return this.http.get<Product>(`${this.apiUrl}/${productId}`);
  }

  // Update a product
  updateProduct(productId: string, product: ProductUpdateDto): Observable<Product> {
    return this.http.put<Product>(`${this.apiUrl}/${productId}`, product);
  }

  // Delete a product
  deleteProduct(productId: string): Observable<{ message: string }> {
    return this.http.delete<{ message: string }>(`${this.apiUrl}/${productId}`);
  }

  // Get all products with pagination
  getAllProducts(page = 0, size = 10, sortBy = 'createdAt', direction = 'desc'): Observable<any> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString())
      .set('sortBy', sortBy)
      .set('direction', direction);
    
    return this.http.get<any>(this.apiUrl, { params });
  }

  // Get active products
  getActiveProducts(): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.apiUrl}/active`);
  }

  // Get products by shop ID
  getProductsByShop(shopId: string): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.apiUrl}/shop/${shopId}`);
  }

  // Get active products by shop ID
  getActiveProductsByShop(shopId: string): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.apiUrl}/shop/${shopId}/active`);
  }

  // Get products by category
  getProductsByCategory(category: string): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.apiUrl}/category/${category}`);
  }

  // Get products by shop and category
  getProductsByShopAndCategory(shopId: string, category: string): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.apiUrl}/shop/${shopId}/category/${category}`);
  }

  // Get featured products
  getFeaturedProducts(): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.apiUrl}/featured`);
  }

  // Get featured products by shop
  getFeaturedProductsByShop(shopId: string): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.apiUrl}/shop/${shopId}/featured`);
  }

  // Search products by name
  searchProductsByName(keyword: string): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.apiUrl}/search/name`, {
      params: new HttpParams().set('keyword', keyword)
    });
  }

  // Search products (name, description, category)
  searchProducts(keyword: string): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.apiUrl}/search`, {
      params: new HttpParams().set('keyword', keyword)
    });
  }

  // Search products within a shop
  searchProductsByShop(shopId: string, keyword: string): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.apiUrl}/shop/${shopId}/search`, {
      params: new HttpParams().set('keyword', keyword)
    });
  }

  // Get products by price range
  getProductsByPriceRange(minPrice: number, maxPrice: number): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.apiUrl}/price-range`, {
      params: new HttpParams()
        .set('minPrice', minPrice.toString())
        .set('maxPrice', maxPrice.toString())
    });
  }

  // Get discounted products
  getProductsWithDiscount(): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.apiUrl}/discounted`);
  }

  // Get discounted products by shop
  getProductsWithDiscountByShop(shopId: string): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.apiUrl}/shop/${shopId}/discounted`);
  }

  // Update product quantity
  updateProductQuantity(productId: string, quantity: number): Observable<Product> {
    return this.http.patch<Product>(`${this.apiUrl}/${productId}/quantity`, null, {
      params: new HttpParams().set('quantity', quantity.toString())
    });
  }

  // Toggle product featured status
  toggleFeaturedStatus(productId: string): Observable<Product> {
    return this.http.patch<Product>(`${this.apiUrl}/${productId}/toggle-featured`, null);
  }

  // Toggle product active status
  toggleActiveStatus(productId: string): Observable<Product> {
    return this.http.patch<Product>(`${this.apiUrl}/${productId}/toggle-active`, null);
  }

  // Get recent products
  getRecentProducts(page = 0, size = 10): Observable<Product[]> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    
    return this.http.get<Product[]>(`${this.apiUrl}/recent`, { params });
  }

  // Get best selling products
  getBestSellingProducts(page = 0, size = 10): Observable<Product[]> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    
    return this.http.get<Product[]>(`${this.apiUrl}/best-selling`, { params });
  }
}